text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,213 @@
+
+Lightweight client SDKs are used to connect to the Gateway for the purpose of invoking transactions.  The gateway will intereact with the rest of the network on behalf of the client eliminating the need for the client application to connect directly to peers and orderers.
+
+The concepts and APIs will be familiar to Fabric programmers who are already using the new programming model.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The Gateway runs as a client process associated with an organisation and requires an identity (issued by the org's CA) in order to interact with the discovery service and to register for events (Deliver Client).  
+
+The Gateway also runs as a gRPC server and exposes the following services to client applications (SDKs):
+
+```
+service Gateway {
+    rpc Endorse(ProposedTransaction) returns (PreparedTransaction) {}
+    rpc Submit(PreparedTransaction) returns (stream Event) {}
+    rpc Evaluate(ProposedTransaction) returns (Result) {}
+}
+```
+
+The client credentials (private key) are never passed to the Gateway.  Client applications are responsible for managing their user keys (optionally using SDK Wallets) and signing the protobuf payloads.
+
+Submitting a transaction is a two step process (performed by the client SDK):
+- __Endorse__
+  - The client creates a ProposedTransaction message, which contains a SignedProposal message as defined in `fabric-protos/peer/proposal.proto` and signed with the user's identity.  The ProposedTransaction message also contains other optional properties, such as the endorsing orgs if the client wants to specify that.
+  - The `Endorse` service is invoked on the Gateway, passing the ProposedTransaction message
+    - The Gateway will determine the endorsement plan for the requested chaincode and forward to the appropriate peers for endorsement. It will return to the client a `PreparedTransaction` message which contains a `Envelope` message as defined in `fabric-protos/common/common.proto`.  It will also contain other information, such as the return value of the chaincode function and the transaction ID so that the client doesn't necessarily need to unpack the `Envelope` payload.
+- __Submit__
+  - The client signs the hash of the Envelope payload using the user's private key and sets the signature field in the `Envelope` in the `PreparedTransaction`.
+  - The `Submit` service is invoked on the Gateway, passing the signed `PreparedTransaction` message.  A stream is opened to return multiple return values.
+    - The Gateway will register transaction event listeners for the given channel/txId.
```

Comment:
I might have missed that, but I was wondering how custom chaincodes events will be supported? My understanding is that Gateway introduced to reduce burden and complexity from the clients where Gateway is to orchestrate transaction endorsement and submission, however, chaincode might result in event that client would be interested in. Are you assume that client will have to register to even service bypassing the new Gateway abstraction?

Reply:
A client-side SDK based on Fabric Gateway would provide pretty much the same API provided by the current client SDKs. So client applications would still be able to listen for chaincode events exactly as they do today.